contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,794 @@
+        _assertZeroOZRoleHolders(config, config.CONFIG_MANAGER_ROLE());
+
+    }
+
+    function _assertInitialProxyImplementations() internal view {
+        if (_withdrawalVault().implementation() != _withdrawalVaultImplementation) revert WrongInitialImplementation(address(_withdrawalVault()));
+        _assertInitialDummyImplementation(_accountingOracle());
+        _assertInitialDummyImplementation(_stakingRouter());
+        _assertInitialDummyImplementation(_validatorsExitBusOracle());
+        _assertInitialDummyImplementation(_withdrawalQueue());
+    }
+
+    function _assertInitialDummyImplementation(IOssifiableProxy proxy) internal view {
+        if (proxy.proxy__getImplementation() != _dummyImplementation) revert WrongInitialImplementation(address(proxy));
+    }
+
+    function _assertSingleOZRoleHolder(IAccessControlEnumerable accessControlled, bytes32 role, address holder) internal view {
+        if (accessControlled.getRoleMemberCount(role) != 1
+         || accessControlled.getRoleMember(role, 0) != holder
+        ) {
+            revert WrongOZAccessControlRoleHolders(address(accessControlled), role);
+        }
+    }
+
+    function _assertTwoOZRoleHolders(IAccessControlEnumerable accessControlled, bytes32 role, address holder1, address holder2) internal view {
+        if (accessControlled.getRoleMemberCount(role) != 2
+         || accessControlled.getRoleMember(role, 0) != holder1
+         || accessControlled.getRoleMember(role, 1) != holder2
+        ) {
+            revert WrongOZAccessControlRoleHolders(address(accessControlled), role);
+        }
+    }
+
+    function _assertCorrectInitialRoleHolders() internal view {
+        _assertSingleOZRoleHolder(_burner(), _burner().REQUEST_BURN_SHARES_ROLE(), address(_lido()));
+
+        _assertZeroOZRoleHolders(_accountingOracle(), DEFAULT_ADMIN_ROLE);
+
+        _assertZeroOZRoleHolders(_stakingRouter(), DEFAULT_ADMIN_ROLE);
+        _assertZeroOZRoleHolders(_stakingRouter(), _stakingRouter().STAKING_MODULE_PAUSE_ROLE());
+        _assertZeroOZRoleHolders(_stakingRouter(), _stakingRouter().STAKING_MODULE_RESUME_ROLE());
+        _assertZeroOZRoleHolders(_stakingRouter(), _stakingRouter().REPORT_EXITED_VALIDATORS_ROLE());
+        _assertZeroOZRoleHolders(_stakingRouter(), _stakingRouter().REPORT_REWARDS_MINTED_ROLE());
+
+        _assertZeroOZRoleHolders(_validatorsExitBusOracle(), DEFAULT_ADMIN_ROLE);
+        _assertZeroOZRoleHolders(_validatorsExitBusOracle(), _validatorsExitBusOracle().PAUSE_ROLE());
+
+        _assertZeroOZRoleHolders(_withdrawalQueue(), DEFAULT_ADMIN_ROLE);
+        _assertZeroOZRoleHolders(_withdrawalQueue(), _withdrawalQueue().PAUSE_ROLE());
+        _assertZeroOZRoleHolders(_withdrawalQueue(), _withdrawalQueue().FINALIZE_ROLE());
+        _assertZeroOZRoleHolders(_withdrawalQueue(), _withdrawalQueue().ORACLE_ROLE());
+    }
+
+    function _assertZeroOZRoleHolders(IAccessControlEnumerable accessControlled, bytes32 role) internal view {
```

Comment:
Move up to other role check functions please

Reply:
done